proposals/0000-warning-pragmas-with-categories.rst

```diff
@@ -0,0 +1,252 @@
+
+This is yet one more feature to implement, although the implementation cost
+should be fairly modest.
+
+Overall this should make the language more accessible to newcomers, as library
+authors will be able to use ``WARNING`` pragmas to discourage certain features
+even if those warnings can be reasonably be disabled in some contexts.
+
+This proposal does not provide a way to disable warnings at specific use sites,
+only at the module level.  In some cases, it would be nice to be able to mark
+individual uses as having been approved and the warning suppressed for that use
+alone, rather than for all uses in the module.
+
+It might be helpful to establish conventions around which categories exist, such
+as ``partial`` for warnings about partial functions.  Moreover it is unclear
+whether every identifier with an attached warning should have a category
+corresponding to the name of the identifier, as with the ``-Wwarnings:head`` and
+``-Wwarnings:tail`` examples.  These issues are currently left to individual
+library authors.
+
+This proposal does not provide a mechanism for organising or namespacing warning
+categories, as they are simply bare identifiers.  Thus if libraries use the
+names of their functions as categories, the names cannot be qualified to
+distinguish definitions from separate modules.
+
+
+Alternatives
+------------
+
+This proposal may be contrasted with `proposal #454
+<https://github.com/ghc-proposals/ghc-proposals/pull/454>`_, which introduces a
+built-in constraint ``Warning`` that can be used for custom warnings along
+similar lines to the existing support for custom type errors.  That proposal
+allows categorisation of warnings in a similar way, and moreover allows
+type-level programming to control the presence and content of warnings, and the
+suppression of warnings at individual use sites.
+
+However, this proposal is simpler, and by keeping the ``WARNING`` annotations as
+separate pragmas rather than requiring them to be part of the types, avoids the
+risk that introducing ``Warning`` constraints may have unexpected effects on
+program semantics.  Assuming this proposal is accepted, it would be fairly
+simple to change the ``Warning`` class to be controlled using the same flags.
+
+`Proposal #454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_ does
+not allow multiple categories to be attached to a single warning, but it seems
+useful to include this feature, because it allows for multiple categories at
+different levels of granularity (potentially including a different category for
+every identifier).
+
+
+Unresolved Questions
+--------------------
+
+GHC currently supports both ``WARNING`` and ``DEPRECATED`` pragmas, which are
+synonymous.  Similarly the flags ``-Wwarnings-deprecations`` and
+``-Wdeprecations`` are synonymous, and there is no ``-Wwarnings`` flag.  This
+seems somewhat inconsistent.  Should we reduce the redundancy by considering
+deprecations one particular category of warnings?
```

Comment:
I'd say so

Reply:
And `-Wwarnings-deprecations` and `-Wdeprecations` simply become synonyms for `-Wwarnings:deprecation` then.